server/src/instant/reactive/store.clj

```diff
@@ -279,89 +256,74 @@
+                                   conn
+                                   [[:db.fn/call clean-stale-subscriptions-tx-data lookup-ref v]
+                                    [:db.fn/call clean-stale-datalog-tx-data]
+                                    [:db.fn/call set-instaql-query-result-tx-data lookup-ref result-hash]])
+        hash-before     (:instaql-query/hash (d/entity (:db-before report) lookup-ref))
+        hash-after      (:instaql-query/hash (d/entity (:db-after report) lookup-ref))
         result-changed? (or (not= hash-before hash-after)
                             (and (nil? hash-before)
                                  (nil? hash-after)))]
-    (assoc res :result-changed? result-changed?)))
+    (assoc report :result-changed? result-changed?)))
 
 ;; ------
 ;; session
 
-(defn get-session [db sess-id]
-  (d/entity db [:session/id sess-id]))
-
-(defn get-session-instaql-queries [db sess-id]
-  (->> (d/q '{:find [?q]
-              :in [$ ?session-id]
-              :where [[?e :instaql-query/session-id ?session-id]
-                      [?e :instaql-query/query ?q]]}
-            db
-            sess-id)
-       (map first)
-       set))
+(defn session-instaql-queries [store app-id sess-id]
+  (let [db @(app-conn store app-id)]
+    (set
+     (for [datom (d/datoms db :avet :instaql-query/session-id sess-id)
+           :let  [ent (d/entity db (:e datom))]]
+       (:instaql-query/query ent)))))
 
-(defn remove-session-queries-tx-data
+(defn- remove-session-queries-tx-data
   "Should be used in a db.fn/call. Returns transactions.
    Retracts queries for the session."
-  [db session-id]
-  (map (fn [{:keys [e]}] [:db/retractEntity e])
-       (d/datoms db :avet :instaql-query/session-id session-id)))
+  [db sess-id]
+  (for [datom (d/datoms db :avet :instaql-query/session-id sess-id)]
+    [:db/retractEntity (:e datom)]))
 
-(defn remove-session-subscriptions-tx-data
+(defn- remove-session-subscriptions-tx-data
   "Should be used in a db.fn/call. Returns transactions.
    Retracts subscriptions for the session."
-  [db session-id]
-  (map (fn [{:keys [e]}] [:db/retractEntity e])
-       (d/datoms db :avet :subscription/session-id session-id)))
-
-(defn remove-session! [conn sess-id]
-  (transact! "store/remove-session!"
-             conn
-             [[:db.fn/retractEntity [:session/id sess-id]]
-              [:db.fn/call remove-session-queries-tx-data sess-id]
-
-              ;; remove subscriptions for session
-              [:db.fn/call remove-session-subscriptions-tx-data sess-id]
-
-              ;; remove datalog-queries that are no longer in use
-              [:db.fn/call clean-stale-datalog-tx-data]]))
-
-;; ------
-;; socket
+  [db sess-id]
+  (for [datom (d/datoms db :avet :subscription/session-id sess-id)]
+    [:db/retractEntity (:e datom)]))
+
+(defn remove-session! [store app-id sess-id]
+  ;; sync so new sessions are not added while we clean up this one
+  (locking (:sessions store)
```

Comment:
Seems a little dangerous if I understand this correctly. This is relying on an implementation detail of datascript to prevent updates to the session db while we're removing the session. But if the `locking` gets removed from datascript, we might not realize that we need to update our code here.

Reply:
We actually have our own locking in `transact!` so should be okay